#11751 Remove ReactorBase._handleSignals

exarkun opened the pull request:
Fixes #11751

exarkun committed:
_FDWaker in general does not need the reactor

So don't give it one.

The "foreign event loop" reactors do need a reactor so preserve theirs.

exarkun committed:
Factor signal handling logic out of ReactorBase (and subclasses)

ReactorBase now sets up a `SignalHandling` object with this behavior.
PosixReactorBase overrides the factory for this object to add SIGCHLD-handling
behavior.

7812431a4f03bfa62ce8d6a4fffaca1793bd7594 provides the original implementation
of the glib bits (thanks glyph).  It looks to me like we don't need to do *as
much* work as that revision does to make glib happy - but all I have to go on
is the test suite.

exarkun committed:
The signal module is always available in this context.

exarkun committed:
If we are on POSIX then process support is always enabled.

Very old Python runtimes might sometimes have been missing some of our process
dependencies but this is no longer supported by the code and no longer much of
an issue in practice.

exarkun committed:
news fragment

exarkun committed:
Python 3.7 compat

exarkun committed:
Make _SocketWaker and _UnixWaker signatures agree

exarkun committed:
Avoid the deprecation warning

It's not clear why this wasn't a problem before, though.

exarkun committed:
Take advantage of MonkeyPatcher already being a context manager

exarkun committed:
remove unused imports

exarkun commented:
The missing coverage is in cfreactor - which CI fails to test *at all* - and wxreactor ... which ... I just don't know what to say.

glyph commented on src/twisted/internet/_glibbase.py:
```diff
@@ -56,11 +57,33 @@ class GlibWaker(_UnixWaker):
         self.reactor._simulate()
 
 
+def _signalGlue():
+    """
+    Integrate glib's wakeup file descriptor usage and our own.
+
+    Python supports only one wakeup file descriptor at a time and both Twisted
+    and glib want to use it.
+
+    This is a context manager that can be wrapped around the whole glib
+    reactor main loop which makes our signal handling work with glib's signal
+    handling.
+    """
+    from gi import _ossighelper as signalGlue  # type: ignore[import]
+
+    patcher = MonkeyPatcher()
+    patcher.addPatch(signalGlue, "_wakeup_fd_is_active", True)
+    return patcher
+
+
```
Beautiful. An excellent refactoring of my attempt. No notes.

glyph commented on src/twisted/internet/_glibbase.py:
```diff
@@ -126,14 +150,18 @@ def __init__(self, glib_module: Any, gtk_module: Any, useGtk: bool = False) -> N
         self._run = self.loop.run
 
-    def _handleSignals(self):
+    def _reallyStartRunning(self):
```
A docstring would be nice here, given the… somewhat unclear distribution of responsibilities in this squirrely mess we are cleaning up

glyph commented on src/twisted/internet/_glibbase.py:
```diff
@@ -126,14 +150,18 @@ def __init__(self, glib_module: Any, gtk_module: Any, useGtk: bool = False) -> N
         # Next, since certain versions of gtk will clobber our signal handler,
         # set all signal handlers again after the event loop has started to
-        # ensure they're *really* set. We don't call this twice so we don't
-        # leak file descriptors created in the SIGCHLD initialization:
-        self.callLater(0, posixbase.PosixReactorBase._handleSignals, self)
+        # ensure they're *really* set.
```
Hmm. Is this still true? I'd love to get a number on "certain versions". Mainly i'm wondering if this is effectively a repeat of similar logic as set up by `_signalGlue`.

glyph replied:
(Idle curiosity here, not necessary to address before landing.)

exarkun replied:
Hm.  Yes, that would be good to know.  The test suite passes on my system with this whole method deleted.  I don't know if that means our test suite is insufficient or if this code is only for obsolete versions of glib or something else.

I'm reluctant to actually delete it since I don't know and since you didn't make this a blocker I'll refrain from doing so _or_ further investigating right here.

exarkun replied:
https://github.com/twisted/twisted/issues/11762 and linked in the code

glyph commented on src/twisted/internet/_signals.py:
```diff
@@ -18,37 +18,45 @@
+
 
-def installHandler(fd):
+def installHandler(fd: int) -> int:
```
love to see a new type annotation

(and I immediately think: should `FileDescriptorInt` be a `NewType`? But I digress)

glyph commented on src/twisted/internet/_signals.py:
```diff
@@ -59,7 +67,8 @@ def installHandler(fd):
         received.
-    @type fd: C{int}
+
+    @return: The file descriptor previously configured for this use.
```
👏

glyph commented on src/twisted/internet/_signals.py:
```diff
@@ -80,6 +89,169 @@ def isDefaultHandler():
+    process-global effects necessary for dealing with signals.
+    """
+
+    def install(self) -> None:
```
This is a state machine, right? Like it should raise an exception if `install` is called twice?

exarkun replied:
I'm not sure if it's useful to look at it that way - but with the mess in the current implementation of `_WithSignalHandling` it's a bit hard to tell.

However, I think about the actual type signature of this method as `IO ()`.  Its purpose in life is to have a side-effect.  So, maybe you want that side-effect to happen twice because something else happened in between and the second call should really just re-install everything?

exarkun replied:
And as detailed in one of my other comments it turns out something, probably trial, manages to drive the process such that `.install(); ... .install()` really happens.

exarkun replied:
https://github.com/twisted/twisted/issues/11763 and linked in the code

glyph commented on src/twisted/internet/_signals.py:
```diff
@@ -80,6 +89,169 @@ def isDefaultHandler():
+        Do nothing for historical reasons.
+        """
+        # TODO Make this do something, someday, because cleaning up your state
+        # is a nice idea.
```
File a ticket, link it here?

glyph commented on src/twisted/internet/_signals.py:
```diff
@@ -80,6 +89,169 @@ def isDefaultHandler():
+        cleanup logic.
+    """
+
+    _delegates: Sequence[SignalHandling]
```
Minor naming nitpick: `_signalHandlings` would avoid introducing a new bit of jargon here.

glyph commented on src/twisted/internet/_signals.py:
```diff
@@ -80,6 +89,169 @@ def isDefaultHandler():
+        Extend the basic signal handling logic to also support handling
+        SIGCHLD to know when to try to reap child processes.
+        """
+        self._childWaker = _SIGCHLDWaker()
```
```suggestion
        assert self._childWaker is None
        self._childWaker = _SIGCHLDWaker()
```

exarkun replied:
It turns out some conch tests manage to make such an assertion fail.  It looks like I didn't exactly replicate the old logic - which tested to see if there was already a waker and skipped making a new one if so (however it call `install` either way).

I think this suggests that trial doesn't "correctly" start/crash the reactor with respect to `_reallyStartRunning`.  The refactoring doesn't change when signal handlers are installed so I think this is as broken in current trunk as it is in this branch - but in both current trunk and this branch (without your suggestion) we just tolerate the brokenness.  I will change the code in the branch to tolerate it exactly the same way as it is tolerated in trunk - by not creating a new waker if we have one already, and then unconditionally installing it (as long as that passes the test suite).

This should be *slightly* better than what's in the branch now since what's in the branch now will leak the clobbered SIGCHLDWaker's file descriptors.

exarkun replied:
https://github.com/twisted/twisted/issues/11763 as mentioned elsewhere

glyph commented on src/twisted/internet/_signals.py:
```diff
@@ -120,9 +292,8 @@ class _SocketWaker(log.Logger):
     disconnected = 0
 
-    def __init__(self, reactor):
+    def __init__(self):
```
while we're in here… how much damage would this do
```suggestion
    def __init__(self) -> None:
```

glyph commented on src/twisted/internet/_signals.py:
```diff
@@ -178,20 +350,19 @@ class _FDWaker(log.Logger):
+    o: int
 
-    def __init__(self, reactor):
+    def __init__(self):
```
```suggestion
    def __init__(self) -> None:
```

glyph commented on src/twisted/internet/base.py:
```diff
@@ -632,32 +646,53 @@ def __init__(self) -> None:
     # Signal handling pieces
     _installSignalHandlers = False
```
```suggestion
    _installSignalHandlers: bool = False
```

glyph commented on src/twisted/internet/base.py:
```diff
@@ -632,32 +646,53 @@ def __init__(self) -> None:
     _installSignalHandlers = False
 
-    def _handleSignals(self) -> None:
+    def _maybeSignalsFactory(self, installSignalHandlers: bool) -> SignalHandling:
```
Naming nitpick: does the `maybe` really add much here?  The active nothing of `_WithoutSignalHandling` means you get a `SignalHandling` no matter what.

glyph commented on src/twisted/internet/cfreactor.py:
```diff
@@ -74,6 +74,10 @@ class _WakerPlus(_UnixWaker):
 
+    def __init__(self, reactor):
+        super().__init__()
+        self.reactor = reactor
```
just looking at the coverage here: 😬 

can we have a separate PR for installing the `macos-platform` extra?  maybe landed before this one?

glyph commented on src/twisted/internet/wxreactor.py:
```diff
@@ -84,7 +84,7 @@ def _installSignalHandlersAgain(self):
         except ImportError:
             return
-        self._handleSignals()
+        self._signals.install()
```
I… think I'm willing to live with the coverage gap here. Is anyone still using wxpython 😬

exarkun replied:
:grimacing:

glyph reviewed (approved):
No blockers. I'd love to get test coverage of cfreactor but if it hasn't been running, debugging the interaction with CI might be a pain.

glyph reviewed (commented):
Apparently I was a bit too optimistic about cfreactor's prospects.  I got some test failures suspiciously close to the reactor core, and then a re-run made them succeed.  Perhaps test_runAfterStop and test_crash have become flaky...?

exarkun committed:
Merge branch 'trunk' into 11751-remove-reactorbase-handlesignals

exarkun committed:
Link to a ticket about uninstall behavior

exarkun committed:
add some notes to GlibReactorBase._reallyStartRunning

exarkun committed:
simpler language

exarkun committed:
Avoid stomping over a real waker

exarkun committed:
a couple easy drive-by annotations

exarkun committed:
improve the SignalHandling helper method name

exarkun committed:
On further consideration, don't require a coherent state

Just deal with the incoherent state in the same way as it was dealt with in
trunk prior to this refactoring.

Some future cleanup or bugfix should fix the reactor and/or trial and/or the
Twisted test suite (and/or something else, who knows!) so that when a
`_ChildSignalHandling` is being installed, it hasn't been installed before.

exarkun committed:
issue link

exarkun committed:
Remove some uncovered lines by removing a branch

Apparently the branch never goes down the bad path so lets just make the bad
path a hard error.

exarkun committed:
Merge branch 'trunk' into 11751-remove-reactorbase-handlesignals

exarkun committed:
Make tearDown consistent with setUp

Also, simplify the whole thing.  There's no reason to call both of these
functions twice.  Just restore the wakeup fd and restore the SIGCHLD handler.
Done.

This didn't fail on my system but it did fail on the gtk2 job on CI.  I hope
this version works.

exarkun committed:
This test really has no business making such demands of global state

Just save the values so we can restore them.  Anything else is unreasonable.

exarkun committed:
Merge branch 'trunk' into 11751-remove-reactorbase-handlesignals
